File: engine/src/codestrata/telemetry/events.py
"""Typed Engine-local telemetry runtime events (Slice 9.1).

Narrow intake model — independent from legacy ``EventName`` / Phase 14.3
payloads and from Platform Community Cloud DTOs.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import Any

from codestrata.telemetry.errors import TelemetryRuntimeError, TelemetryRuntimeErrorCode


class RuntimeEventType(StrEnum):
    """Initial runtime event vocabulary aligned toward future cloud mapping."""

    APPLICATION_STARTED = "application_started"
    APPLICATION_COMPLETED = "application_completed"
    FEATURE_INVOKED = "feature_invoked"
    FEATURE_COMPLETED = "feature_completed"
    OPERATION_FAILED = "operation_failed"


class OsFamily(StrEnum):
    LINUX = "linux"
    MACOS = "macos"
    WINDOWS = "windows"
    OTHER = "other"


class ArchFamily(StrEnum):
    X86_64 = "x86_64"
    ARM64 = "arm64"
    OTHER = "other"


class Lifecycle(StrEnum):
    START = "start"
    COMPLETE = "complete"
    FAIL = "fail"


class ResultCategory(StrEnum):
    SUCCESS = "success"
    FAILURE = "failure"
    CANCELLED = "cancelled"
    UNKNOWN = "unknown"


class DurationBucket(StrEnum):
    LT_1S = "lt_1s"
    S_1_10 = "s_1_10"
    S_10_60 = "s_10_60"
    M_1_5 = "m_1_5"
    GT_5M = "gt_5m"


class OperationCategory(StrEnum):
    ASSESS = "assess"
    REPORT = "report"
    OTHER = "other"


class FailureCategory(StrEnum):
    VALIDATION = "validation"
    TIMEOUT = "timeout"
    UNAVAILABLE = "unavailable"
    INTERNAL = "internal"
    UNKNOWN = "unknown"


APPROVED_RUNTIME_EVENT_TYPES: frozenset[str] = frozenset(
    item.value for item in RuntimeEventType
)


@dataclass(frozen=True, slots=True)
class RuntimeTelemetryEvent:
    """Narrow typed runtime event intake — no repository/path/credential fields."""

    event_type: RuntimeEventType
    client_name: str = "codestrata_cli"
    cli_version: str | None = None
    os_family: OsFamily | None = None
    arch_family: ArchFamily | None = None
    lifecycle: Lifecycle | None = None
    result: ResultCategory | None = None
    duration_bucket: DurationBucket | None = None
    operation_category: OperationCategory | None = None
    enabled_assessment_heads: tuple[str, ...] = ()
    offline_mode: bool | None = None
    ai_used: bool | None = None
    failure_category: FailureCategory | None = None

    def __post_init__(self) -> None:
        if self.client_name != "codestrata_cli":
            raise TelemetryRuntimeError(TelemetryRuntimeErrorCode.UNSAFE_VALUE)
        if not isinstance(self.event_type, RuntimeEventType):
            raise TelemetryRuntimeError(TelemetryRuntimeErrorCode.VALIDATION_FAILED)
        heads = tuple(sorted({str(item) for item in self.enabled_assessment_heads if item}))
        for head in heads:
            if "/" in head or "\\" in head or ".." in head or len(head) > 32:
                raise TelemetryRuntimeError(TelemetryRuntimeErrorCode.UNSAFE_VALUE)
        object.__setattr__(self, "enabled_assessment_heads", heads)
        if self.cli_version is not None:
            version = self.cli_version.strip()
            if not version or len(version) > 32 or "/" in version or "\\" in version:
                raise TelemetryRuntimeError(TelemetryRuntimeErrorCode.UNSAFE_VALUE)
            object.__setattr__(self, "cli_version", version)

    def to_intake_dict(self) -> dict[str, Any]:
        """Stable dict of present fields for projection (not a public dump)."""

        payload: dict[str, Any] = {
            "client_name": self.client_name,
            "event_type": self.event_type.value,
        }
        if self.cli_version is not None:
            payload["cli_version"] = self.cli_version
        if self.os_family is not None:
            payload["os_family"] = self.os_family.value
        if self.arch_family is not None:
            payload["arch_family"] = self.arch_family.value
        if self.lifecycle is not None:
            payload["lifecycle"] = self.lifecycle.value
        if self.result is not None:
            payload["result"] = self.result.value
        if self.duration_bucket is not None:
            payload["duration_bucket"] = self.duration_bucket.value
        if self.operation_category is not None:
            payload["operation_category"] = self.operation_category.value
        if self.enabled_assessment_heads:
            payload["enabled_assessment_heads"] = list(self.enabled_assessment_heads)
        if self.offline_mode is not None:
            payload["offline_mode"] = self.offline_mode
        if self.ai_used is not None:
            payload["ai_used"] = self.ai_used
        if self.failure_category is not None:
            payload["failure_category"] = self.failure_category.value
        return {key: payload[key] for key in sorted(payload)}


__all__ = [
    "APPROVED_RUNTIME_EVENT_TYPES",
    "ArchFamily",
    "DurationBucket",
    "FailureCategory",
    "Lifecycle",
    "OperationCategory",
    "OsFamily",
    "ResultCategory",
    "RuntimeEventType",
    "RuntimeTelemetryEvent",
]

File: vscode-plugin/src/telemetry/events.ts
/**
 * Bounded VS Code telemetry event model (Slice 9.13).
 * Schema: community-vscode-telemetry-event-schema:1.0
 */

export const VSCODE_CLIENT_NAME = "vscode_extension" as const;

export type VsCodeRuntimeEventType =
  | "feature_invoked"
  | "feature_completed"
  | "operation_failed";

export type VsCodeLifecycle = "start" | "complete" | "fail";
export type VsCodeResultCategory = "success" | "failure" | "cancelled" | "unknown";
export type VsCodeOperationCategory = "assess" | "report" | "other";
export type VsCodeOsFamily = "linux" | "macos" | "windows" | "other";
export type VsCodeDurationBucket =
  | "lt_1s"
  | "s_1_10"
  | "s_10_60"
  | "m_1_5"
  | "gt_5m";

export const APPROVED_EVENT_TYPES: readonly VsCodeRuntimeEventType[] = [
  "feature_invoked",
  "feature_completed",
  "operation_failed",
];

export const APPROVED_FIELD_NAMES = [
  "event_type",
  "client_name",
  "extension_version",
  "os_family",
  "lifecycle",
  "result",
  "duration_bucket",
  "operation_category",
  "offline_mode",
  "ai_used",
  "schema_version",
  "runtime_policy_version",
] as const;

export type VsCodeRuntimeTelemetryEvent = {
  readonly eventType: VsCodeRuntimeEventType;
  readonly clientName: typeof VSCODE_CLIENT_NAME;
  readonly extensionVersion?: string;
  readonly osFamily?: VsCodeOsFamily;
  readonly lifecycle?: VsCodeLifecycle;
  readonly result?: VsCodeResultCategory;
  readonly durationBucket?: VsCodeDurationBucket;
  readonly operationCategory?: VsCodeOperationCategory;
  readonly offlineMode?: boolean;
  readonly aiUsed?: boolean;
};

export function createAssessInvokedEvent(options: {
  aiUsed: boolean;
  extensionVersion?: string;
}): VsCodeRuntimeTelemetryEvent {
  return {
    eventType: "feature_invoked",
    clientName: VSCODE_CLIENT_NAME,
    extensionVersion: options.extensionVersion,
    operationCategory: "assess",
    lifecycle: "start",
    aiUsed: options.aiUsed,
  };
}

export function createAssessCompletedEvent(options: {
  aiUsed: boolean;
  success: boolean;
  cancelled?: boolean;
  extensionVersion?: string;
  durationBucket?: VsCodeDurationBucket;
}): VsCodeRuntimeTelemetryEvent {
  if (options.cancelled) {
    return {
      eventType: "feature_completed",
      clientName: VSCODE_CLIENT_NAME,
      extensionVersion: options.extensionVersion,
      operationCategory: "assess",
      lifecycle: "complete",
      result: "cancelled",
      aiUsed: options.aiUsed,
      durationBucket: options.durationBucket,
    };
  }
  return {
    eventType: options.success ? "feature_completed" : "operation_failed",
    clientName: VSCODE_CLIENT_NAME,
    extensionVersion: options.extensionVersion,
    operationCategory: "assess",
    lifecycle: options.success ? "complete" : "fail",
    result: options.success ? "success" : "failure",
    aiUsed: options.aiUsed,
    durationBucket: options.durationBucket,
  };
}

export function eventToIntakeDict(
  event: VsCodeRuntimeTelemetryEvent
): Record<string, unknown> {
  const payload: Record<string, unknown> = {
    client_name: event.clientName,
    event_type: event.eventType,
  };
  if (event.extensionVersion !== undefined) {
    payload.extension_version = event.extensionVersion;
  }
  if (event.osFamily !== undefined) {
    payload.os_family = event.osFamily;
  }
  if (event.lifecycle !== undefined) {
    payload.lifecycle = event.lifecycle;
  }
  if (event.result !== undefined) {
    payload.result = event.result;
  }
  if (event.durationBucket !== undefined) {
    payload.duration_bucket = event.durationBucket;
  }
  if (event.operationCategory !== undefined) {
    payload.operation_category = event.operationCategory;
  }
  if (event.offlineMode !== undefined) {
    payload.offline_mode = event.offlineMode;
  }
  if (event.aiUsed !== undefined) {
    payload.ai_used = event.aiUsed;
  }
  return payload;
}
